### PR TITLE
riscv64 builds

### PR DESCRIPTION
**Pull request by selfisekai**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -192,10 +201,36 @@ jobs:
...
+          echo "
+          https://dl-cdn.alpinelinux.org/alpine/edge/main
+          https://dl-cdn.alpinelinux.org/alpine/edge/community
+          " > /etc/apk/repositories
+
       - name: Install sysroot
         run: |
           cd dart-sdk/sdk
-          ./build/linux/alpine_sysroot_scripts/install-sysroot.sh
+          dart_to_alpine_arch() {
```
Comment (ntkme) [inline .github/workflows/build.yml]: While it is nice trying to install only what's absolutely needed, installing extra sysroots only takes a few seconds. Comparing to that the whole job takes about 2-3 hours on GitHub Action, it not worth saving a few seconds.

Therefore, I would prefer just keep installing all variants of sysroot for simplicity. This is also what the official dart sdk does for its debian sysroots that all of them get installed regardless of what is going to be compiled.

As for `riscv64`, you can just add an extra step to install it for now. Once `riscv64` lands in alpine stable, we can send the patch to upstream.

Comment (selfisekai) [reply .github/workflows/build.yml]: ok, PR simplified then.

the patch to dart (https://dart-review.googlesource.com/c/sdk/+/296901) makes the script accept a list of architectures or by default download the previous list of architectures (without riscv64) so I'm planning to submit that as it is - I don't think riscv64 is getting shipped anytime soon in alpine, 3.18 currently being built definitely will not

Comment (ntkme) [reply .github/workflows/build.yml]: The patch looks good.

**Comment by ntkme** [comment]
As for getting edge repo. I think a better way is to conditionally use edge image:

``` yaml
      image: docker.io/library/alpine:${{ matrix.target-arch == 'riscv64' && 'edge' || 'latest'}}
```

This way we don't mix stable and edge in `/etc/apk/repositories`

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -145,7 +152,8 @@ jobs:
     runs-on: ubuntu-latest
 
     container:
-      image: docker.io/library/alpine
+      # riscv64 is currently alpine edge only
+      image: docker.io/library/alpine:${{ matrix.target-arch == 'riscv64' && 'edge' || 'latest'}}
```
Comment (ntkme) [inline .github/workflows/build.yml]: Looks like I missed a cosmetic whitespace in my previous comment:

```suggestion
      image: docker.io/library/alpine:${{ matrix.target-arch == 'riscv64' && 'edge' || 'latest' }}
```

Comment (selfisekai) [reply .github/workflows/build.yml]: I didn't notice that either tbh

**Comment by ntkme** [comment]
LGTM. As for https://github.com/dart-lang/sdk/commit/0f549e2069ff78fedf43a9781ce5b80625b9d470, are you going to submit it to upstream?

**Comment by selfisekai** [comment]
submitted here: https://dart-review.googlesource.com/c/sdk/+/296901

**Comment by ntkme** [comment]
Once your patch get merged into upstream, let's update the commit sha for the patch to the official one. I will merge this PR after that. Thanks for your work!

**Comment by ntkme** [comment]
This has landed. Here is the first release for riscv64: https://github.com/dart-musl/dart/releases/tag/3.1.0-189.0.dev

**Pull request closed without merging**